### PR TITLE
fix: concurrent map write cause by concurrent invocation of function …

### PR DESCRIPTION
**Pull request by liushv0**
…DeepCopy (#2167)

### Type of change:

- [x] Bugfix
- [ ] New feature provided
- [ ] Improve performance
- [ ] Backport patches
- [ ] Documentation
- [ ] Refactor
- [ ] Chore
- [ ] CI/CD or Tests

### Pre-submission checklist:

- [ ] Did you explain what problem does this PR solve? Or what new features have been added?
- [ ] Have you added corresponding test cases?
- [ ] Have you modified the corresponding document?
- [ ] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix-ingress-controller#community) first**

### CODE REVIEW

**Comment by Revolyssup** [comment]
The struct on which unmarshal is being performed is created with new() everytime when DeepCopy is called. Which means the data which is being written is not shared between function calls, so how can there be concurrent writes? I am trying to understand that right now.

**Comment by Revolyssup** [comment]
Because we have the same kind of implementation as there are in kubernetes repos and nowhere do they use a lock 
https://github.com/search?q=org%3Akubernetes%20DeepCopy(&type=code

**Comment by Revolyssup** [comment]
@liushv0 Can you answer the above questions to help me review this PR?

**Comment by liushv0** [comment]
image


sorry for late  :( 

apisix-route sync goroutine（gr1）will get test-ar from MemDB，
apisix-plugin-config sync goroutine （gr2 ）will get test-apc from MemDB，
test-apc has be refered by test-ar。


now  gr1 begin to sync CR test-apc，func `PluginConfig.DeepCopyInto` will be called，in this function，`Plugins#DeepCopyInto` be called。
image



at the same time， gr2 begin to sync CR test-apc，func  `Route#DeepCopyInto`  will be called，in this function，`Plugins#DeepCopyInto`  also be called。 
image


Plugins in test-apc and Plugins in test-ar  pointer to same instance，so two goroutine concurrent  marshal  this Plugins instance，and panic

**Review by Revolyssup — APPROVED** [review]

**Comment by Revolyssup** [comment]
@liushv0 I think there is no point in having locks as well as DeepCopy. I think this line need to be removed https://github.com/apache/apisix-ingress-controller/blob/58aab33bc6d215c6a3f826daedbd5a5d85ca6c23/pkg/types/apisix/v1/zz_generated.deepcopy.go#L336. This line may have been added when this file was autto generated.

**Comment by Revolyssup** [comment]
@liushv0 Are you available to make changes here?

File: pkg/apisix/cache/memdb.go [hunk 1]
```diff
@@ -113,6 +127,8 @@ func (c *dbCache) GetUpstream(id string) (*v1.Upstream, error) {
 	if err != nil {
 		return nil, err
 	}
+	c.upstreamLock.Lock()
+	defer c.upstreamLock.Unlock()
```
Comment (Revolyssup) [inline pkg/apisix/cache/memdb.go:131]: @liushv0 I think there is no point in having locks as well as DeepCopy. I think this line need to be removed https://github.com/apache/apisix-ingress-controller/blob/58aab33bc6d215c6a3f826daedbd5a5d85ca6c23/pkg/types/apisix/v1/zz_generated.deepcopy.go#L336. This line may have been added when this file was autto generated.

**Review by Revolyssup — CHANGES_REQUESTED** [review]

File: pkg/apisix/cache/memdb.go (hunk 1 above)
Comment (liushv0) [reply pkg/apisix/cache/memdb.go:131]: I guess remove  this line is effective，and less change，
but honesty，I am not exactly understand  the intention of this line，set *out = *in for what。
other side，this line  is generated by code-gen tools，maybe somebody will rerun  code-gen tools，and overwrite the  modification。
that is why I choose use  locks。

**Comment by Revolyssup** [comment]
@liushv0 Can you modify this PR to remove locks and instead apply this suggestion?https://github.com/apache/apisix-ingress-controller/pull/2168#discussion_r1527877936

**Comment by lkad** [comment]
i have meet same bug ,has anyone can fix  it ?

**Comment by lkad** [comment]
@liushv0  你这段代码在自己环境中使用了吗，我目前也碰到同样的问题, 烦请有空告知下，非常感谢.

**Pull request closed without merging**